main.go

```diff
@@ -3,36 +3,120 @@
 package main
 
 import (
+	"errors"
 	"flag"
+	"os"
+	"os/signal"
 	"strconv"
+	"syscall"
 
 	"github.com/docker/go-plugins-helpers/volume"
 	"github.com/golang/glog"
+	"github.com/mellanox-senior-design/docker-volume-rdma/db"
 	"github.com/mellanox-senior-design/docker-volume-rdma/drivers"
 )
 
 // Port to launch service on.
 var httpPort int
+var volumeDatabaseDriver string
+var volumeDatabasePath string
+var volumeDatabaseHost string
+var volumeDatabaseUsername string
+var volumeDatabasePassword string
+var volumeDatabaseSchema string
 
-// Configure and start the docker volume plugin server.
-func main() {
-
+func init() {
 	// Configure application flags.
 	flag.IntVar(&httpPort, "port", 8080, "tcp/ip port to serve volume driver on")
 
+	flag.StringVar(&volumeDatabaseDriver, "db", "sqlite", "set the database backend used to store volume metadata: [sqlite, mysql, in-memory]")
+	flag.StringVar(&volumeDatabasePath, "dbpath", "", "set the database storage path")
+
+	flag.StringVar(&volumeDatabaseHost, "dbhost", "", "set the database host (default is '' or localhost:3306)")
+	flag.StringVar(&volumeDatabaseUsername, "dbuser", "", "set the database username (default is root)")
+	flag.StringVar(&volumeDatabasePassword, "dbpass", "", "set the database password (optional)")
+	flag.StringVar(&volumeDatabaseSchema, "dbschema", "", "set the database schema (required)")
+}
+
+// Configure and start the docker volume plugin server.
+func main() {
+
 	// Parse flags as glog needs the flags to be solidified before starting.
 	flag.Parse()
 
 	// Convert port to string, and print startup message.
 	port := strconv.Itoa(httpPort)
-	glog.Info("Running! http://localhost:" + port)
 
 	// Create and begin serving volume driver on tcp/ip port, httpPort.
+	vd, dberr := GetDatabaseConnection()
+	if dberr != nil {
+		glog.Fatal(dberr)
+	}
+
+	// Configure Storage Controller
 	sc := drivers.NewGlusterStorageController()
-	driver := drivers.NewRDMAVolumeDriver(sc)
+
+	// Print startup message and start server
+	glog.Info("Connecting to services ...")
+	driver := drivers.NewRDMAVolumeDriver(sc, vd)
+	driver.Connect()
+	defer driver.Disconnect()
+
+	// Handle SIGINT gracefully
+	c := make(chan os.Signal, 2)
+	signal.Notify(c, os.Interrupt, syscall.SIGINT)
+	go func() {
+		<-c
+		glog.Info("Exiting ...")
+		driver.Disconnect()
+		os.Exit(1)
+	}()
+
+	glog.Info("Running! http://localhost:" + port)
 	h := volume.NewHandler(driver)
 	err := h.ServeTCP("test_volume", ":"+port, nil)
 
 	// Log any error that may have occurred.
 	glog.Fatal(err)
 }
+
+// GetDatabaseConnection returns the database connection that was requested on the command line.
+func GetDatabaseConnection() (db.VolumeDatabase, error) {
+	switch volumeDatabaseDriver {
+	case "in-memory":
+		validateDatabaseFlags(true, true, false, false, false, false)
```

Comment:
After some brief searching I found this: https://groups.google.com/forum/#!topic/golang-nuts/oWeeqCJfRzk

They basically made a config struct and passed the struct after initialization around. Performance-wise this would be identical, but reading-wise it would be far more obvious what's going on. Let's do that instead